### PR TITLE
Merge branch 'refactor' to 'dev'

### PR DESCRIPTION
**Pull request by hugefiver**
refactor source to adjust go 1.18 feature like generic

### CODE REVIEW

**Review by hugefiver — COMMENTED** [review]
reviewed

File: util/result.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
+package util
+
+import (
+	"reflect"
+)
+
+// Result is a type contains result.
+type Result[T any] struct {
```
Comment (icceey) [inline util/result.go:8]: 这是干啥的呀

File: util/range.go [hunk 2]
```diff
@@ -1,33 +1,41 @@
...
-func (r RangeInt) IsEmpty() bool {
+func (r Range[T]) IsEmpty() bool {
 	return r.max <= r.min
 }
 
-// NewRangeInt return a RangeInt object.
-func NewRangeInt(min, max int) RangeInt {
-	return RangeInt{
+// NewRangeInt return a Range[Int] object.
+func NewRangeInt[T Interger](min, max T) Range[T] {
```
Comment (icceey) [inline util/range.go:20]: 直接用`NewRange`吧，别`NewRangeInt`了

File: util/range.go [hunk 3]
```diff
@@ -1,33 +1,41 @@
...
 		max: max,
 		min: min,
 	}
 }
 
-// NewEmptyRangeInt return a empty RangeInt object.
-func NewEmptyRangeInt() RangeInt {
-	return RangeInt{
+// NewEmptyRangeInt return a empty Range[Int] object.
+func NewEmptyRangeInt[T Interger]() Range[T] {
```
Comment (icceey) [inline util/range.go:28]: 不如改成`func NewEmptyRange[T Ordered]() Range[T]`

**Review by icceey — COMMENTED** [review]
1 problem found
2 optimizations found

File: util/result.go (hunk 1 above)
Comment (hugefiver) [reply util/result.go:8]: 不到啊

**Review by hugefiver — COMMENTED** [review]
reviewed

File: util/range.go [hunk 4]
```diff
@@ -1,33 +1,86 @@
...
+	max T
+	min T
+	t   IntervalType
 }
 
 // Cover return if pattern `x` is in range.
-func (r RangeInt) Cover(x int) bool {
-	return x > r.min && x < r.max
+func (r Range[T]) Cover(x T) bool {
+	return (x > r.min || (r.t&leftClosed == leftClosed) && x == r.min) &&
```
Comment (icceey) [inline util/range.go:41]: 感觉封装一下`r.IsLeftClosed`会更好，不过这次先这样吧

**Review by icceey — APPROVED** [review]

**Pull request merged**